Shopify/cli
# Migrate functions with legacy IDs to extension IDs when a value exists on remote

### WHY are these changes introduced?
#gsd:30370

Functions are moving onto unified app deployments. This means they have a new ID on the server. Since functions are being backfilled on the server, this means the CLI has to somehow connect to them.

### WHAT is this pull request doing?

Automatically migrate functions with legacy IDs to extension IDs whenever the unified app beta flag is set and a matching extension exists on the server. The matching is done by checking for the legacy ID in the server-side extension configuration.

### How to test your changes?

Theres a few server-side PRs that I need to ship to make this work (which should happen soon) but theres no problem with merging this before those. It just wont do anything.

🟡 You'll need these branches: partners: `js.deploy-functions-on-ef`, shopify: `js.automerge-ef`, script-service: `js.deploy-to-ef-3`. They will be responsible for mirroring extensions and functions on the server anytime you deploy using a CLI version without the unified app deployments beta.

- **Without** the unified app deployments beta, deploy an app with a function.
- See that you have ULID formatted IDs
    - <img width="717" alt="image" src="https://github.com/Shopify/cli/assets/28009669/a265bec3-6259-4dcc-8fdd-8d54528f84e1">
- Set the unified app deployments beta on your app
- Deploy again
- See that you have UUID formatted IDs
    - <img width="717" alt="image" src="https://github.com/Shopify/cli/assets/28009669/d0b03041-0865-4332-bbba-45c8eb6dc119">



### Questions
- Any concerns with this approach? We do other similar things so I'm assuming its fine. There likely will have to be manual steps for apps that are deployed in a CI environment but we can document that when we're ready for release.

### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
